### PR TITLE
Add java

### PR DESCRIPTION
**Pull request by MatMoore**
This is my first draft of the Java port, which uses maven to manage dependencies.

Differences from the python version:

- I didn't create an entry point for the program (a class with a main() method) because it seems like you can do the exercise with just classes and tests
- Added a getBalance() method because java doesn't have properties

Before merging this I'm going to see if I can bundle gradle as well (to give the candidate a choice) and check that other people besides me can build and run this.

### CODE REVIEW

File: java/src/test/java/OysterCardTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  @Test
+  public void testSwipeIn() {
+    OysterCard card = new OysterCard();
+    card.topUp(1000);
+
+    card.touchIn("TUBE", "Earl's Court");
+
+    assertEquals(card.getBalance(), 680);
```
Comment (exonian) [inline java/src/test/java/OysterCardTest.java]: Might these assertions be better the other way around? The failure messages I'm seeing are like this:
```org.opentest4j.AssertionFailedError: expected: <-1> but was: <680>```

3 options from here:
1) it's valuable to have these slightly confusing, cos that's life; 2) it's valuable to have them as clear as possible; 3) it really doesn't matter

**Review by exonian — COMMENTED** [review]
Nice! I brew installed maven and it was utterly plain sailing from there! Lovely stuff

One comment, about a thing which probably confused me more than it might someone actually doing the test...

File: java/src/test/java/OysterCardTest.java (hunk 1 above)
Comment (MatMoore) [reply java/src/test/java/OysterCardTest.java]: Good point! I didn't notice this. I think it's valuable to have them as clear as possible - interviews are stressful enough already so I don't think we should make the test harder that it needs to be

Comment (exonian) [reply java/src/test/java/OysterCardTest.java]: 100% agree, but then overthought it 😁

**Comment by exonian** [comment]
Thanks @MatMoore! 🎉

**Pull request merged**